rectools/visuals/metrics_app.py

```diff
@@ -0,0 +1,158 @@
+        plotly_kwargs : tp.Optional[tp.Dict[str, tp.Any]], optional, default None
+            Additional arguments for customizing the `px.scatter` plot.
+
+        Returns
+        -------
+        MetricsApp
+            An instance of `MetricsApp`, providing interactive Jupyter widget for metric visualization.
+        """
+        return cls(models_metrics, show_legend, auto_display, plotly_kwargs)
+
+    def _validate_input_data(self) -> None:
+        if not isinstance(self.models_metrics, pd.DataFrame):
+            raise ValueError("Incorrect input type. `metrics_data` should be a DataFrame")
+        if Columns.Split not in self.models_metrics.columns:
+            raise KeyError("Missing 'Split' column in `metrics_data` DataFrame")
+        if Columns.Model not in self.models_metrics.columns:
+            raise KeyError("Missing 'Model' column in `metrics_data` DataFrame")
+        if len(self.models_metrics.columns) < 3:
+            raise KeyError("`metrics_data` DataFrame assumed to have at least one metric column")
+
+    @property
+    def metric_names(self) -> tp.List[str]:
+        """List of metric column names from the `models_metrics`."""
+        non_metric_columns = {Columns.Split, Columns.Model}
+        return [col for col in self.models_metrics.columns if col not in non_metric_columns]
+
+    @property
+    def fold_ids(self) -> tp.List[int]:
+        """Sorted list of unique fold identifiers from the `models_metrics`."""
+        return sorted(self.models_metrics[Columns.Split].unique())
+
+    @property
+    def n_folds(self) -> int:
+        """Total number of unique folds available in the `models_metrics`."""
+        return len(self.fold_ids)
+
+    def _make_chart_data(self, fold_number: int) -> pd.DataFrame:
+        return self.models_metrics[self.models_metrics[Columns.Split] == fold_number].drop(columns=Columns.Split)
+
+    def _make_chart_data_avg(self) -> pd.DataFrame:
+        return self.models_metrics.drop(columns=Columns.Split).groupby(Columns.Model, sort=False).mean().reset_index()
+
+    def _toggle_fold_number_visibility(self, fold: widgets.Dropdown, use_avg: widgets.Checkbox) -> None:
+        fold.layout.visibility = "hidden" if use_avg.value else "visible"
+
+    def _update_chart(
+        self,
+        scatter_chart: widgets.Output,
+        metric_x: widgets.Dropdown,
+        metric_y: widgets.Dropdown,
+        use_avg: widgets.Checkbox,
+        fold: widgets.Dropdown,
+    ) -> None:
+        with scatter_chart:
+            scatter_chart.clear_output(wait=True)
+            chart_data = self._make_chart_data_avg() if use_avg.value else self._make_chart_data(fold.value)
+            self.fig = px.scatter(
+                data_frame=chart_data,
+                x=metric_x.value,
+                y=metric_y.value,
+                color=Columns.Model,
+                width=self.plotly_kwargs.get("width", DEFAULT_WIDTH),
```

Comment:
This will cause an error like `multiple values for argument 'width'` if `width` in the `self.plotly_kwargs`

Reply:
Sorry, I don't get your idea. How dictionary could possibly have multiple values for one key? Anyway, this part is also outdated now.

Reply:
Fixed, thank you.